### PR TITLE
starter: demonstrate io and other seastar techniques

### PR DESCRIPTION
**Pull request by dotnwat**

### CODE REVIEW

File: main.cc [hunk 1]
```diff
@@ -1,71 +1,240 @@
...
-        ss << "msg: \"" << _msg << "\" from core "
-           << seastar::this_shard_id();
-        co_await seastar::sleep(std::chrono::seconds(seastar::this_shard_id()));
-        lg.info("Processed speak request");
-        co_return ss.str();
+    seastar::future<> start() {
+        file_ = co_await seastar::open_file_dma(
+          path_.string(), seastar::open_flags::ro);
+
+        const auto stat = co_await file_.stat();
```
Comment (rockwotj) [inline main.cc]: nit: could also just call `file_.size()`.

File: main.cc [hunk 2]
```diff
@@ -1,71 +1,240 @@
...
+
+        /*
+         * limit input size to be a multiple of page size for simpler math.
+         */
+        if (stat.st_size % page_size != 0) {
+            if (seastar::this_shard_id() == 0) {
```
Comment (rockwotj) [inline main.cc]: Why do this check only on shard 0?

File: main.cc [hunk 3]
```diff
@@ -1,71 +1,240 @@
...
+          start_page_,
+          end_page_);
+
+        /*
+         * invokes `run()` in the background. in order to be able to synchronize
+         * with the background fiber running it is started under a `gate` which
+         * can be used to wait until the background fiber finishes, which is
+         * done in `stop()`.
+         */
+        (void)seastar::with_gate(gate_, [this] { return run(); });
```
Comment (rockwotj) [inline main.cc]: nit: `std::ignore = seastar::with_gate(...);`

**Review by rockwotj — APPROVED** [review]
Very nice. A couple of mega nits, but LGTM.

File: main.cc (hunk 3 above)
Comment (dotnwat) [reply main.cc]: fixed

File: main.cc (hunk 2 above)
Comment (dotnwat) [reply main.cc]: fixed

File: main.cc (hunk 1 above)
Comment (dotnwat) [reply main.cc]: fixed

**Review by rockwotj — APPROVED** [review]

**Comment by dotnwat** [comment]
i haven't looked at it in more detail, but it is _wild_ how slowly the debug build runs

**Comment by nvartolomei** [comment]
Have missed this but came to say the new example is great!

**Pull request merged**